### PR TITLE
KAFKA-9620: Do not throw in the middle of consumer user callbacks

### PR DESCRIPTION
**Pull request by abbccdda**
One way of fixing it forward.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ static void wipeStateStores(final Logger log, final ProcessorStateManager stateM
             Utils.delete(stateMgr.baseDir());
         } catch (final IOException fatalException) {
             // since it is only called under dirty close, we always swallow the exception
-            log.warn("Failed to wiping state stores for task {}", stateMgr.taskId());
+            log.warn("Failed to wiping state stores for task {} due to {}", stateMgr.taskId(), fatalException);
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:87]: Add a logging to log the IO exception.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 2]
```diff
@@ -293,12 +301,11 @@ boolean tryToCompleteRestoration() {
      * @throws TaskMigratedException if the task producer got fenced (EOS only)
      */
     void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
-        final Set<TaskId> revokedTasks = new HashSet<>();
         final Set<TopicPartition> remainingPartitions = new HashSet<>(revokedPartitions);
 
         for (final Task task : tasks.values()) {
             if (remainingPartitions.containsAll(task.inputPartitions())) {
-                revokedTasks.add(task.id());
+                task.suspend();
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:312]: Simplify the logic here as we no longer throws before task suspend

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:312]: Not sure I follow: we can still throw from inside `suspend` right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 3]
```diff
@@ -198,12 +198,19 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
         }
 
         if (!taskCloseExceptions.isEmpty()) {
+            for (final Map.Entry<TaskId, RuntimeException> entry : taskCloseExceptions.entrySet()) {
+                if (!(entry.getValue() instanceof TaskMigratedException)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:202]: I think we should use `instanceof KafkaException` as well here:  if it is any inheritance of KafkaException, we should not wrap it.

**Review by guozhangwang — COMMENTED** [review]
Thanks @abbccdda , made a pass.

Could you also add the unit test coverage in this PR?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 2 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:312]: What I mean is that we do not throw if remaining partitions are not empty anymore. We could directly suspend a task here.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:312]: I see.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 3 above)
Comment (abbccdda) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:202]: Sounds good

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -741,6 +733,85 @@ public void shouldHaveRemainingPartitionsUncleared() {
...
+
+        final TaskMigratedException thrown = assertThrows(TaskMigratedException.class,
+            () -> taskManager.handleAssignment(emptyMap(), emptyMap()));
+        // The task map orders tasks based on topic group id and partition, so here
+        // t1 should always be the first.
+        assertThat(thrown.getMessage(), equalTo("t1 close exception; it means all tasks belonging to this thread should be migrated."));
+    }
+
+    @Test
+    public void shouldThrowRuntimeExceptionWhenEncounteredUnknownExceptionDuringTaskClose() {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:762]: Thank you for the added tests!

**Comment by guozhangwang** [comment]
test this please

**Comment by abbccdda** [comment]
@guozhangwang should be good to review again

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 5]
```diff
@@ -338,14 +353,11 @@ public void testManyInFlightAsyncCommitsWithCoordinatorDisconnect() {
...
             Map<TopicPartition, OffsetAndMetadata> offsets = singletonMap(tp, new OffsetAndMetadata(i));
-            coordinator.commitOffsetsAsync(offsets, new OffsetCommitCallback() {
-                @Override
-                public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
-                    responses.incrementAndGet();
-                    Throwable cause = exception.getCause();
-                    assertTrue("Unexpected exception cause type: " + (cause == null ? null : cause.getClass()),
-                            cause instanceof DisconnectException);
-                }
+            coordinator.commitOffsetsAsync(offsets, (offsets1, exception) -> {
```
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:356]: just a side cleanup

**Comment by abbccdda** [comment]
Got 1/2 green build

**Comment by abbccdda** [comment]
The failure was due to non related flaky test:
kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition


```Error Message
org.scalatest.exceptions.TestFailedException: Timed out before consuming expected 2700 records. The number consumed was 1620.
Stacktrace
org.scalatest.exceptions.TestFailedException: Timed out before consuming expected 2700 records. The number consumed was 1620.
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1389)
	at org.scalatest.Assertions.fail(Assertions.scala:1091)
	at org.scalatest.Assertions.fail$(Assertions.scala:1087)
	at org.scalatest.Assertions$.fail(Assertions.scala:1389)```

**Comment by guozhangwang** [comment]
test this please

**Comment by guozhangwang** [comment]
test this please

**Comment by ijuma** [comment]
Does this affect AK 2.5?

**Comment by guozhangwang** [comment]
From our soak observations it only affects 2.5 not now; there's a major tech debt cleanup that only get into trunk so this issue is likely a regression in trunk only.

**Pull request merged**